### ISSUE TITLE
consider working together on ruby-sys

### ISSUE THREAD

**Issue by steveklabnik** [rustbridge/helix#2]
Hey yinz,

Psyched to see this become an actual thing!

For another "Rust in Ruby" project, I recently extracted https://github.com/steveklabnik/ruby-sys . From looking at your vendored libmrisys, it looks like it's got almost all the same stuff. Maybe we can make the two compatible, so you don't have to worry about the bindings anymore?

**Comment by eternaleye** [comment]
There's also https://github.com/dylanmckay/mri-sys which is used by https://github.com/dylanmckay/rurust (embedding Ruby in Rust, rather than extending Ruby using Rust).

Along similar lines to rurust, there's https://github.com/anima-engine/mrusty (embedding mruby, rather than mri) - if helix ever grows support for embedding, it may be worth looking at how mrusty handles injecting things into the runtime (it injects virtual files, which can be loaded with `require` from the embedded Ruby interpreter).

**Comment by chancancode** [comment]
👍 in principle

There are a few challenges. The main one is that a lot of "functions" in CRuby are actually macros. Obviously, we don't have access to those at runtime, so we need a glue layer to "re-export" them as functions/named symbols so they can be called from Rust.

How this works now is that we ship a small C-extension called [helix_runtime](https://github.com/rustbridge/helix/blob/master/ruby/ext/helix_runtime/native/helix_runtime.c) that we build on the end-user's machine using the standard C compilation toolchain.

If we were to merge the code we also need to move that extension and keep that in-sync with the Rust binding.

Also 👍 on figuring out the grand unification between supporting both C Ruby and mruby. (@hone02 will be interested in this.)

**Comment by steveklabnik** [comment]
> so we need a glue layer to "re-export" them as functions/named symbols so they can be called from Rust.

Yeah, this is tough. The aproach ruru has taken is to just port them, but that can be a bit of a problem, if things diverge...

**Comment by malept** [comment]
One would hope that if things diverge, it would happen in a major version bump of Ruby, [but I guess that's not necessarily the case](https://www.ruby-lang.org/en/news/2013/12/21/ruby-version-policy-changes-with-2-1-0/).

**Comment by hone** [comment]
I'm definitely interested, but my impression is that the mruby/mri API is fairly different.

**Comment by wagenet** [comment]
Seems like we could still use ruby-sys, if we wanted to. I'm not sure if it gains us a huge benefit right now, but any sort of standardization sounds beneficial. One thing work considering is how this would interact with our efforts for safe function calls in #64. @steveklabnik

**Comment by steveklabnik** [comment]
> I'm not sure if it gains us a huge benefit right now, but any sort of standardization sounds beneficial. 

I feel like it benefits *users* more than each project, that is, it's more about the interop than anything else.

Stuff has stagnated for me a bit lately, but I'm still down. I've been wanting to look into https://github.com/steveklabnik/ruby-sys/issues/14 sometime soon, but my spare open source time has been.... small lately 😦

**Comment by wagenet** [comment]
@steveklabnik :+1: Seems worth giving it a shot.
